README.md
Gitab Cookbook
=========================

This cookbook will setup a Gitlab instance. It will use the Gitlab
Omnibus package for installation.

Requirements
------------

# Cookbooks

* none

Attributes
----------

Please see the supplied attributes file for common defaults and how to
change them. Other attributes not listed can be found on the official
omnibus package (e.g. how to write `the gitlab.rb` file).

Usage
-----

Add `recipe[gitlab]` in your nodes run list.

Data Bag
========

* `gitlab`

This container holds all information to setup a new instance.

* `ldap`

The ldap item holds optional LDAP server information, see supplied
encrypted data bag using knife solo data bag for details.

* `certificate`

The certificate item holds an optional SSL certificate which will be
installed in Nginx. If you set your `external_url` to https, ensure you
have a data bag ready holding your certificate. See supplied encrypted
data bag using knife solo data bag for details.

Data Bag Location
-----------------

* `test/integration/data_bags` for Kitchen

Data Bag Format
---------------

* Use `knife solo data bag` to create and edit local data bags

## License & Authors
- Author:: Sebastian Grewe <sebgrewe@bigpoint.net>

```text
Copyright:: 2014 Bigpoint GmbH

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.



CHANGELOG.md
## 1.1.1:

* Added additional file to rotation, default to 14 days

## 1.1.0:

* Added logrotate as dependency and into configure step

## 1.0.0:

* Initial release of rabbitmq-service

- - -
Check the [Markdown Syntax Guide](http://daringfireball.net/projects/markdown/syntax) for help with Markdown.

The [Github Flavored Markdown page](http://github.github.com/github-flavored-markdown/) describes the differences between markdown on github and standard markdown.


spec/spec_helper.rb
# encoding: utf-8
require 'chefspec'
require 'chefspec/berkshelf'
require 'chefspec/deprecations'

@platforms = { 'ubuntu' => ['12.04', '14.04'], 'debian' => ['7.6'], 'centos' => ['6.5', '7.0'] }

at_exit { ChefSpec::Coverage.report! }


spec/default_spec.rb
# encoding: utf-8
require 'spec_helper'

@platforms.each do |platform, versions|
  versions.each do |version|
    describe "The gitlab-omni::default #{platform} version #{version} recipe" do
      before do
        allow(Chef::EncryptedDataBagItem).to receive(:load).with(
          'gitlab', 'certificate'
        ).and_return('')
        allow(Chef::EncryptedDataBagItem).to receive(:load).with('gitlab', 'ldap').and_return('')
        @chef_run = ChefSpec::SoloRunner.new(platform: platform, version: version)
        @chef_run.converge 'gitlab-omni::default'
      end

      %w(gitlab-omni::_definitions
         gitlab-omni::_install_from_omnibus
         gitlab-omni::_install_certificates
         gitlab-omni::_configure_gitlab).each do |rcp|
        it 'should include recipe ' + rcp do
          expect(@chef_run).to include_recipe rcp
        end
      end
    end
  end
end


spec/_definitions_spec.rb
# encoding: utf-8
require 'spec_helper'

@platforms.each do |platform, versions|
  versions.each do |version|
    describe "The gitlab-omni::_definitions #{platform} version #{version} recipe" do
      before do
        @chef_run = ChefSpec::SoloRunner.new(platform: platform,
                                             version: version)
        @chef_run.converge 'gitlab-omni::_definitions'
      end

      it 'bash[gitlab-setup] should do nothing' do
        resource = @chef_run.bash('gitlab-setup')
        expect(resource).to do_nothing
      end

      # Doesn't work as expected yet
      it 'should subscribe to other resources'
    end
  end
end


recipes/_configure_gitlab.rb
# encoding: utf-8
# Load LDAP information from encrypted data bag for better security
data_bag_name = node['gitlab-omni']['ldap_data_bag_name']
data_bag_item = node['gitlab-omni']['ldap_data_bag_item']
begin
  db_ldap = Chef::EncryptedDataBagItem.load(
    data_bag_name,
    data_bag_item
  ) if node['gitlab-omni']['rails']['config']['ldap_enabled']
rescue
  raise 'You have https enabled but I was unable to load your data bag' \
    "(#{data_bag_name}, #{data_bag_item}) containing the information"
end

template '/etc/gitlab/gitlab.rb' do
  source 'gitlab.rb.erb'
  user 'root'
  group 'root'
  mode '600'
  variables('ldap_host' => db_ldap['host'],
            'ldap_port' => db_ldap['port'],
            'ldap_uid' => db_ldap['uid'],
            'ldap_method' => db_ldap['method'],
            'ldap_bind_dn' => db_ldap['bind_dn'],
            'ldap_password' => db_ldap['password'],
            'ldap_allow_username_or_email_login' => db_ldap['allow_username_or_email_login'],
            'ldap_base' => db_ldap['base'],
            'ldap_user_filter' => db_ldap['user_filter']
           ) if node['gitlab-omni']['rails']['config']['ldap_enabled']
end

# Setup a backup cron
cron 'gitlab-backup' do
  minute node['gitlab-omni']['backup']['minute']
  hour node['gitlab-omni']['backup']['hour']
  weekday node['gitlab-omni']['backup']['weekday']
  command '/opt/gitlab/bin/gitlab-rake gitlab:backup:create'
  only_if { node['gitlab-omni']['backup']['enabled'] }
end

# Setup logrotation
logrotate_app 'gitlab' do
  cookbook 'logrotate'
  path node['gitlab-omni']['logrotate']['path']
  frequency node['gitlab-omni']['logrotate']['frequency']
  options node['gitlab-omni']['logrotate']['options']
  rotate node['gitlab-omni']['logrotate']['rotate']
end


recipes/_install_certificates.rb
# encoding: utf-8
# Load SSL certificates from encrypted data bag
if node['gitlab-omni']['config']['external_url'].start_with?('https')
  data_bag_name = node['gitlab-omni']['certificate_data_bag_name']
  data_bag_item = node['gitlab-omni']['certificate_data_bag_item']
  begin
    db_nginx = Chef::EncryptedDataBagItem.load(data_bag_name, data_bag_item)
  rescue
    raise 'You have https enabled but I was unable to load your data bag' \
          "(#{data_bag_name}, #{data_bag_item}) containing the information"
  end

  # Create folder for cert/key files
  dirs = Set.new
  [node['gitlab-omni']['nginx']['config']['ssl_certificate'],
   node['gitlab-omni']['nginx']['config']['ssl_certificate_key']].each do |file|
    dirs << File.dirname(file)
  end

  dirs.each do |dir|
    directory dir do
      user 'root'
      group 'root'
      mode '700'
      action :create
    end
  end

  file node['gitlab-omni']['nginx']['config']['ssl_certificate'] do
    owner 'root'
    group 'root'
    mode '0600'
    content "#{db_nginx['cert']}\n#{db_nginx['ca']}\n"
  end

  file node['gitlab-omni']['nginx']['config']['ssl_certificate_key'] do
    owner 'root'
    group 'root'
    mode '0600'
    content db_nginx['certkey']
  end
end


attributes/default.rb
# Package information
# Load package information depending on node configuration
default['gitlab-omni']['omnibus']['url'] = value_for_platform(
  'debian' => {
    'default' => 'https://downloads-packages.s3.amazonaws.com/debian-7.6/gitlab_7.3.2-omnibus-1_amd64.deb'
  },
  'ubuntu' => {
    '12.04' => 'https://downloads-packages.s3.amazonaws.com/ubuntu-12.04/gitlab_7.3.2-omnibus-1_amd64.deb',
    '14.04' => 'https://downloads-packages.s3.amazonaws.com/ubuntu-14.04/gitlab_7.3.2-omnibus-1_amd64.deb'
  },
  'centos' => {
    '7.0' => 'https://downloads-packages.s3.amazonaws.com/centos-7.0.1406/gitlab-7.3.2_omnibus-1.el7.x86_64.rpm',
    '7.0.1406' => 'https://downloads-packages.s3.amazonaws.com/centos-7.0.1406/gitlab-7.3.2_omnibus-1.el7.x86_64.rpm',
    '6.5' => 'https://downloads-packages.s3.amazonaws.com/centos-6.5/gitlab-7.3.2_omnibus-1.el6.x86_64.rpm'
  }
)
default['gitlab-omni']['omnibus']['checksum'] = value_for_platform(
  'debian' => {
    'default' => '8cadae4bee9aba0ffdb1025d1075c1a3fcd973ffb3b292b6592a5ab6179dcd16'
  },
  'ubuntu' => {
    '12.04' => '36b7a1c449c94da27c807391e08c371a0385c50202596750a5244930f2c1b1e0',
    '14.04' => 'b0551e4f8e18dce0c6848c9e2498b57a00904b5710b4665a000941d0faff6c4f'
  },
  'centos' => {
    '7.0' => 'c3de7aa6979997190014660ac1bb4848879e8bdcdd797e8231d07446feb1b3db',
    '7.0.1406' => 'c3de7aa6979997190014660ac1bb4848879e8bdcdd797e8231d07446feb1b3db',
    '6.5' => 'df57fea7cf3025f515424525a7f49c0c4a4d70f34fb41df7435d5972f35db70a'
  }
)

# Log rotation
default['gitlab-omni']['logrotate']['path'] = ['/var/log/gitlab/gitlab-rails/*.log',
                                               '/var/log/gitlab/nginx/*.log',
                                               '/var/log/gitlab/gitlab-shell/gitlab-shell.log',
                                               '/var/log/gitlab/unicorn/*.log']
default['gitlab-omni']['logrotate']['frequency'] = 'daily'
default['gitlab-omni']['logrotate']['rotate'] = 14
default['gitlab-omni']['logrotate']['options'] = ['missingok', 'compress', 'notifempty', 'copytruncate']



# Some data bag information in case we want to modify this on the fly
default['gitlab-omni']['ldap_data_bag_name'] = 'gitlab'
default['gitlab-omni']['ldap_data_bag_item'] = 'ldap'
default['gitlab-omni']['certificate_data_bag_name'] = 'gitlab'
default['gitlab-omni']['certificate_data_bag_item'] = 'certificate'

# This is the nginx configuration block to configure the gitlab.rb which generates the vhost for gitlab, see
# https://gitlab.com/gitlab-org/omnibus-gitlab/blob/master/files/gitlab-cookbooks/gitlab/templates/default/nginx-gitlab-http.conf.erb
# This certificate is loaded from the gitlab nginx encrypted databag
default['gitlab-omni']['nginx']['config']['redirect_http_to_https'] = true
default['gitlab-omni']['nginx']['config']['ssl_certificate'] = '/etc/gitlab/ssl/' + node['fqdn'] + '.crt'
default['gitlab-omni']['nginx']['config']['ssl_certificate_key'] = '/etc/gitlab/ssl/' + node['fqdn'] + '.key'

# Default backup cycles, each day at 2 AM system time
default['gitlab-omni']['backup']['enabled'] = true
default['gitlab-omni']['backup']['minute'] = '0'
default['gitlab-omni']['backup']['hour'] = '2'
default['gitlab-omni']['backup']['weekday'] = '*'

# Some static configuration variables used in the /etc/gitlab/gitlab.rb
default['gitlab-omni']['config']['external_url'] = 'https://' + node['fqdn']
default['gitlab-omni']['config']['git_data_dir'] = '/var/opt/gitlab/git-data'

# This is the gitlab configuration block to configure the gitlab.rb which generates the gitlab.yml, see
# https://gitlab.com/gitlab-org/omnibus-gitlab/blob/master/files/gitlab-cookbooks/gitlab/templates/default/gitlab.yml.erb
# LDAP information has been moved into the gitlab ldap encrypted databag
default['gitlab-omni']['rails']['config']['ldap_enabled'] = false
default['gitlab-omni']['rails']['config']['gitlab_default_projects_features_visibility_level'] = 'internal'
default['gitlab-omni']['rails']['config']['gravatar_enabled'] = true
default['gitlab-omni']['rails']['config']['gitlab_email_from'] = 'gitlab@' + node['domain']
default['gitlab-omni']['rails']['config']['smtp_enable'] = false
default['gitlab-omni']['rails']['config']['smtp_address'] = 'localhost'
default['gitlab-omni']['rails']['config']['smtp_port'] = 25
default['gitlab-omni']['rails']['config']['smtp_domain'] = node['domain']
default['gitlab-omni']['rails']['config']['smtp_authentication'] = 'login'
default['gitlab-omni']['rails']['config']['smtp_user_name'] = nil
default['gitlab-omni']['rails']['config']['smtp_password'] = nil
default['gitlab-omni']['rails']['config']['smtp_enable_starttls_auto'] = true
